vtm/src/org/oscim/tiling/source/mapfile/MapDatabase.java

```diff
@@ -717,35 +719,57 @@ private boolean processWayDataBlock(MapElement e, boolean doubleDeltaEncoding, b
             /* each way node consists of latitude and longitude */
             int len = numWayNodes * 2;
 
-            wayLengths[coordinateBlock] = decodeWayNodes(doubleDeltaEncoding,
-                    e, len, isLine);
-
+            GeoPoint[] waySegment = null;
             if (wayCoordinates != null) {
                 // create the array which will store the current way segment
-                GeoPoint[] waySegment = new GeoPoint[e.getNumPoints()];
-                for (int i = 0; i < e.getNumPoints(); i++)
-                    waySegment[i] = new GeoPoint(e.getPointY(i) / 1E6, e.getPointX(i) / 1E6);
+                waySegment = new GeoPoint[e.getNumPoints()];
+            }
+
+            wayLengths[coordinateBlock] = decodeWayNodes(doubleDeltaEncoding,
+                    e, len, isLine, coordinateBlock == 0 ? labelPosition : null, waySegment);
+
+            if (wayCoordinates != null)
+            {
+                if (waySegment.length < wayLengths[coordinateBlock] / 2)
+                {
+                    // shortn waySegment array if points were dropped
+                    waySegment = Arrays.copyOf(waySegment, wayLengths[coordinateBlock] / 2);
```

Comment:
No very nice for performance / memory to create arrays continuously.
What is the exact need for that check and array copy?

Reply:
**From previous function TileProjection.project():**
points of a way get removed when:

- /* drop small distance intermediate nodes */
- /* remove identical start/end point */
so ---> After reading in a way it can have less points then stored in mapfile.

Need to allocate waySegment before filling it, but it is not possible to know how many points will get dropped before reading it. So i needed to shorten the waySegment after.

Reply:
waySegment is only used in reverseGeocoding.
I assume removing mentioned points is done because they are not needed for rendering.

Maybe we could just include those points in waySegment, so we do not need to shorten the array after filling it.